integration/utils/try.go

```diff
@@ -1,50 +1,209 @@
+	if timeout <= 0 {
+		panic("timeout must be larger than zero")
+	}
+
+	interval := time.Duration(math.Ceil(float64(timeout) / 15.0))
+	if interval > maxInterval {
+		interval = maxInterval
+	}
+
+	ci := os.Getenv("CI")
+	if len(ci) > 0 {
+		log.Println("Activate CI multiplier:", CITimeoutMultiplier)
+		timeout = time.Duration(float64(timeout) * CITimeoutMultiplier)
+	}
+
+	var err error
+	if err = operation(); err == nil {
+		return nil
+	}
+
+	stopTime := time.Now().Add(timeout)
+
+	for {
+		if time.Now().After(stopTime) {
+			fmt.Println("-")
+			return fmt.Errorf("try operation failed: %s", err)
+		}
+
+		// FIXME refactor this part
+		select {
+		case <-time.Tick(interval):
+			fmt.Print("*")
+			if err = operation(); err == nil {
+				fmt.Println("+")
+				return nil
+			}
+		}
+	}
+}
+
+// Sleep pauses the current goroutine for at least the duration d.
+// Use only when use an other Try[...] functions is not possible.
+func Sleep(d time.Duration) {
+	ci := os.Getenv("CI")
+	if len(ci) > 0 {
+		log.Println("Activate CI multiplier:", CITimeoutMultiplier)
+		d = time.Duration(float64(d) * CITimeoutMultiplier)
+	}
+	time.Sleep(d)
+}
+
 // Condition is a retry condition function.
 // It receives a response, and returns an error
 // if the response failed the condition.
 type Condition func(*http.Response) error
 
-// ErrorIfStatusCodeIsNot returns a retry condition function.
-// The condition returns an error
-// if the given response's status code is not the given HTTP status code.
-func ErrorIfStatusCodeIsNot(status int) Condition {
+// ComposeCondition compose several `Conditions`.
+func ComposeCondition(conditions ...Condition) Condition {
+	return func(resp *http.Response) error {
+		for _, cond := range conditions {
+			err := cond(resp)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+// BodyContains returns a retry condition function.
+// The condition returns an error if the request body does not contain the given
+// string.
+func BodyContains(s string) Condition {
+	return func(res *http.Response) error {
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read response body: %s", err)
+		}
+
+		if !strings.Contains(string(body), s) {
+			return fmt.Errorf("could not find '%s' in body '%s'", s, string(body))
+		}
+		return nil
+	}
+}
+
+// UntilStatusCodeIs returns a retry condition function.
+// The condition returns an error if the given response's status code is not the
+// given HTTP status code.
+func UntilStatusCodeIs(status int) Condition {
 	return func(res *http.Response) error {
 		if res.StatusCode != status {
-			return errors.New("Bad status. Got: " + res.Status + ", expected:" + strconv.Itoa(status))
+			return fmt.Errorf("got status code %d, wanted %d", res.StatusCode, status)
 		}
 		return nil
 	}
 }
+
+func doGetResponse(url string, timeout time.Duration, condition Condition) (*http.Response, error) {
```

Comment:
This one is not required, let's just pass the http `method` to `doResponse` and use only `doRespose`